api/types/src/transaction.rs

```diff
@@ -1119,45 +1124,226 @@ impl VerifyInput for Secp256k1EcdsaSignature {
     }
 }
 
-impl TryFrom<Secp256k1EcdsaSignature> for TransactionAuthenticator {
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Union)]
+#[serde(tag = "type", rename_all = "snake_case")]
+#[oai(one_of, discriminator_name = "type", rename_all = "snake_case")]
+pub enum Signature {
+    Ed25519(HexEncodedBytes),
+    Secp256k1Ecdsa(HexEncodedBytes),
+}
+
+impl TryFrom<Signature> for AnySignature {
     type Error = anyhow::Error;
 
-    fn try_from(value: Secp256k1EcdsaSignature) -> Result<Self, Self::Error> {
-        let Secp256k1EcdsaSignature {
-            public_key,
-            signature,
-        } = value;
-        Ok(TransactionAuthenticator::secp256k1_ecdsa(
-            public_key
-                .inner()
-                .try_into()
-                .context("Failed to parse given public_key bytes as a Secp256k1EcdsaPublicKey")?,
-            signature
-                .inner()
-                .try_into()
-                .context("Failed to parse given signature as a Secp256k1EcdsaSignature")?,
-        ))
+    fn try_from(signature: Signature) -> Result<Self, Self::Error> {
+        Ok(match signature {
+            Signature::Ed25519(s) => AnySignature::ed25519(s.inner().try_into()?),
+            Signature::Secp256k1Ecdsa(s) => AnySignature::secp256k1_ecdsa(s.inner().try_into()?),
+        })
     }
 }
 
-impl TryFrom<Secp256k1EcdsaSignature> for AccountAuthenticator {
+impl From<AnySignature> for Signature {
+    fn from(signature: AnySignature) -> Self {
+        match signature {
+            AnySignature::Ed25519 { signature } => {
+                Signature::Ed25519(signature.to_bytes().to_vec().into())
+            },
+            AnySignature::Secp256k1Ecdsa { signature } => {
+                Signature::Secp256k1Ecdsa(signature.to_bytes().to_vec().into())
+            },
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Union)]
+#[serde(tag = "type", rename_all = "snake_case")]
+#[oai(one_of, discriminator_name = "type", rename_all = "snake_case")]
+pub enum PublicKey {
```

Comment:
@davidiw, what is the point of this struct?

Is it meant to be a serialized-form of `AnyPublicKey`?

Reply:
Yes, the Json representation... this is what sucks, we end up with the AnyPublicKey represented 3 times...